[Popping community - step 1] opixxx

opixxx opened the pull request:
일반적인 커뮤니티에서의 유저, 게시판, 게시글 기능을 간단히 구현

유저
- 로그인, 회원가입 기능 구현
- 세션을 이용하여 인증 구현, 인가 기능은 나중에 스프링 인터셉터로 구현할 계획
- 비밀번호 암호화를 스프링 시큐리티 bcryptpasswordencoder를 사용하지 않고 구현할 방법을 고민 중

게시판
- 게시판 CRU 구현, 삭제같은 경우는 게시글과 관계 떄문에 게시판 삭제를 어떻게 처리할지 고민해봐야 함

게시글
-Step 2에서 게시글 작성을 다룰 예정이라 바뀌는 부분이 많아 push 하지는 않았음

opixxx committed:
chore: properties 파일 gitingore에 추가

opixxx committed:
chore: 의존성 추가

data jpa, validation, mysql 의존성 추가함

opixxx committed:
feat: BaseEntity 기능 추가

엔티티 생성, 변경될 때 시간을 추적하기 위해 Auditing 기능 추가

opixxx committed:
feat: BaseEntity 기능 추가

엔티티 생성, 변경될 때 시간을 추적하기 위해 Auditing 기능 추가

opixxx committed:
chore: Security 의존성 추가

BCryptPasswordEncoder 를 사용해서 비밀번호를 암호화 하기 위해 추가

opixxx committed:
feat: 로그인, 회원가입, 로그아웃 기능 구현

Session 방식을 사용하여 유저 로그인, 회원가입, 로그아웃 기능 구현

opixxx committed:
chore: h2 의존성 추가

opixxx committed:
chore: yml ignore 추가

opixxx committed:
test: UserRepository 테스트 코드 작성

opixxx committed:
test: 회원가입, 로그인 테스트 코드 작성

opixxx committed:
test: UserController 테스트 코드 작성

opixxx committed:
chore: properties -> yml 변경

opixxx committed:
chore: 의존성 수정

opixxx committed:
refactor: Auditing 기능 Config로 분리

테스트 코드 작성 시 Auditing을 생성하지 않아 테스트 코드 실패함 Config 파일로 분리해서 해결

opixxx committed:
refactor: 시큐리티 config 삭제

opixxx committed:
style: DTO 클래스 명 변경

opixxx committed:
refactor: User 관련 코드 수정

테이블명 user -> users
register() 메서드 반환값 변경 UserRegisterResponse

opixxx committed:
refactor: 생성자 접근 제어자 public -> private

opixxx committed:
refactor: @Transactional 추가

opixxx committed:
feat: Board CRUD 기능 구현

게시판(Board) 생성, 조회, 수정기능 구현

opixxx committed:
test: BoardRepository 테스트코드 작성

opixxx committed:
test: BoardService 테스트 코드 작성

opixxx committed:
test: BoardController 테스트 코드 작성

opixxx committed:
test: 테스트 환경 통합하기

Dr-KoKo commented on src/main/java/org/c4marathon/assignment/global/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,30 @@
+
+@Slf4j
+@RestControllerAdvice
+public class GlobalExceptionHandler {
```
```suggestion
public class GlobalExceptionHandler extends ResponseEntityExceptionHandler{
```

`ResponseEntityExceptionHandler`를 사용하는 방안도 고민해보시면 좋을 것 같아요.
많은 보일러 플레이트를 줄일 수 있고 에러 메시지에 관한 RFC 9457에 대응하여 에러를 핸들링 할 수 있습니다.

PaikMyeongGyu commented on src/main/java/org/c4marathon/assignment/board/presentation/BoardController.java:
```diff
@@ -0,0 +1,42 @@
+    @PostMapping("/v1/board")
+    public ResponseEntity<BoardResponse> createBoard(@Valid @RequestBody BoardCreateRequest request) {
+
+        BoardResponse board = boardService.createBoard(request.toServiceDto());
```
게시글 생성에서 꼭 응답이 필요한지 잘모르겠습니다! + Service 계층 메서드가 Controller에서의 응답을 알고 있어서 차라리 생성만 하고 ResponseEntity<Void>로 놓는 게 전 낫다고 생각합니다.

opixxx replied:
void로 수정했습니다.

Dr-KoKo commented on src/test/java/org/c4marathon/assignment/board/service/BoardServiceTest.java:
```diff
@@ -0,0 +1,115 @@
+
+import static org.assertj.core.api.Assertions.*;
+
+class BoardServiceTest extends IntegrationTestSupport {
```
서비스 레이어 테스트를 통합테스트로 구축하신 이유가 있을까요?

PaikMyeongGyu commented on src/main/java/org/c4marathon/assignment/board/service/BoardService.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.Optional;
+
+@RequiredArgsConstructor
+@Transactional
```
클래스 레벨에 @Transactoinal 다는 것보단 메서드 단위로 다는 게 낫지 않을까? 싶습니다.

Dr-KoKo commented on src/main/java/org/c4marathon/assignment/user/service/UserService.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+
+    public User login(UserLoginServiceRequest loginDto) {
```
User(Entity)를 트랜잭션 바깥으로 return 했을 때 발생할 수 있는 문제점도 고민해봐야 할 것 같아요.
lazy fatch로 인해 예외가 발생할 수도 있고, 이를 방지하려면 OSIV를 사용하게 될 것 같네요.

opixxx replied:
userId를 반환하도록 수정했습니다.

Dr-KoKo commented on src/main/java/org/c4marathon/assignment/user/service/UserService.java:
```diff
@@ -0,0 +1,60 @@
+        return user;
+    }
+
+    private boolean validateNicknameDuplicate(String nickname) {
+        return userRepository.existsByNickname(nickname);
+    }
+
+    private boolean validateEmailDuplicate(String email) {
+        return userRepository.existsByEmail(email);
+    }
```
검증을 위해 Query를 두 번 날라갈 것 같아요.
User를 한번 불러오고 User를 통해 검증하는 방법이 더 효율적일 것 같은 생각이 듭니다.

Dr-KoKo reviewed (commented):
궁금한 점 리뷰로 남겼습니다.
수고하셨어요~

Dr-KoKo committed:
docs: PR 템플릿 생성

Dr-KoKo committed:
docs: sonar 업데이트

Dr-KoKo committed:
docs: sonar 업데이트

Dr-KoKo committed:
docs: jacoco 의존성 삭제

Dr-KoKo committed:
docs: jacoco 의존성 복원

opixxx committed:
style: 사용안하는 import문 제거

opixxx committed:
fix: 세션에 저장하는 유저 정보 수정

opixxx committed:
refactor: User 엔티티 트랜잭션 반환 수정 -> userId를 반환

opixxx committed:
feat: 닉네임에 공백을 검증하는 로직 추가

opixxx committed:
refactor: 예외 처리 기능 수정

- 에외를 던질 때 ErrorCode를 명시적으로 넣지 않고, 예외 생성자 안에서 해당하는 ErrorCode를 설정해서 넣음
- 예외에 중복으로 들어가는 ErrorCode는 없기 때문에 그렇게 수정함

opixxx committed:
refactor: 게시판 생성 반환값 수정

opixxx committed:
fix: ErrorCode 수정

opixxx committed:
test: 바뀐 로직에 대한 테스트 코드 수정
